fix(ngcc): fix UMD inline exports and re-exports

gkalpak opened the pull request:
Several fixes regarding exports in UMD.
See individual commits for more info.

gkalpak committed:
fix(ngcc): correctly handle inline exports in UMD

This fix was part of a broader `ngtsc`/`ngcc` fix in 02bab8cf9 (see
there for details). In 02bab8cf9, the fix was only applied to
`CommonJsReflectionHost`, but it is equally applicable to
`UmdReflectionHost`. Later in #34254, the fix was partially ported to
`UmdReflectionHost` by fixing the `extractUmdReexports()` method.

This commit fully fixes `ngcc`'s handling of inline exports for code in
UMD format.

gkalpak committed:
fix(ngcc): recognize re-exports with `require()` calls in UMD

Previously, `UmdReflectionHost` would only recognize re-exports of the
form `__export(someIdentifier)` and not `__export(require('...'))`.
However, it is possible in some UMD variations to have the latter format
as well. See discussion in https://github.com/angular/angular/pull/34254/files#r359515373

This commit adds support for re-export of the form
`__export(require('...'))` in UMD.

gkalpak committed:
refactor(ngcc): share code between `CommonJsReflectionHost` and `UmdReflectionHost`

While different, CommonJS and UMD have a lot in common regarding the
their exports are constructed. Therefore, there was some code
duplication between `CommonJsReflectionHost` and `UmdReflectionHost`.

This commit extracts some of the common bits into a separate file as
helpers to allow reusing the code in both `ReflectionHost`s.

mary-poppins commented:
You can preview 4a27b8d at https://pr34512-4a27b8d.ngbuilds.io/.

gkalpak committed:
fixup! refactor(ngcc): share code between `CommonJsReflectionHost` and `UmdReflectionHost`

gkalpak committed:
refactor(ngcc): avoid returning the same top-level helper calls multiple times

This change should not have any impact on the code's behavior (based on
how the function is currently used), but it will avoid unnecessary work.

gkalpak committed:
refactor(ngcc): use a special map for memoizing expensive-to-compute values

Previously, in cases were values were expensive to compute and would be
used multiple times, a combination of a regular `Map` and a helper
function (`getOrDefault()`) was used to ensure values were only computed
once.

This commit uses a special `Map`-like structure to compute and memoize
such expensive values without the need to a helper function.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -54,47 +55,42 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
-      if (i.parameter === importParameter) {
-        return i.path;
+      for (const i of imports) {
+        this.umdImportPaths.set(i.parameter, i.path);
```
Arguably this line is no longer needed since the `getOrDefault()` function will write the return value of this factory to the map.
That being said, this is a minor optimisation, since it adds additional imports, that precede the desired one, to the map, speeding up future lookups on those imports. But in that case perhaps we should complete the iteration over all the imports and add them all the first time we get here?

gkalpak replied:
This looked intentional, so I left it as is.
Adding all imports to the map sounds good to me (we will almost certainly use them anyway).

gkalpak replied:
Done.

petebacondarwin commented on packages/compiler-cli/ngcc/src/utils.ts:
```diff
@@ -64,6 +64,13 @@ export function findAll<T>(node: ts.Node, test: (node: ts.Node) => node is ts.No
   }
 }
 
+export function getOrDefault<K, V>(map: Map<K, V>, key: K, factory: (key: K) => V): V {
```
I know that I originally named this function but I wonder if "default" is the correct term here?
What this function is really doing is memoizing a complex factory result...

Perhaps we could create a MemoizedMap class that looks like:

```ts
class MemoizedMap<K,V> extends Map {
  constructor(private factory: (key: T) => V) {}
  get() {
    if (!this.has(key)) {
      this.set(key, factory(key));
    }
    return this.get(key) !;
  }
}
```

gkalpak replied:
Done.

(Extending `Map` causes errors in transpiled ES5 code, so I used a `Map` internally instead. I also went with `FactoryMap`, because that felt more appropriate.)

petebacondarwin reviewed (approved):
Couple of nits - nice tidying up.

gkalpak commented:
@petebacondarwin, thx for the review. PTAL at the last 3 commits, which are new.

mary-poppins commented:
You can preview a7bce14 at https://pr34512-a7bce14.ngbuilds.io/.

gkalpak commented:
Rebased on master.

mary-poppins commented:
You can preview c310fae at https://pr34512-c310fae.ngbuilds.io/.

petebacondarwin commented:
CI does not seem happy...

gkalpak committed:
fix(ngcc): avoid error due to circular dependency in `EsmDependencyHost`

Previously, there was circular dependency between `ngcc/src/utils.ts`,
`ngcc/src/dependencies/dependency_host.ts` and
`ngcc/src/dependencies/esm_dependency_host.ts`. More specifically,
`utils.ts` would [import from `esm_dependency_host.ts`][1], which would
[import from `dependency_host.ts`][2], which would in turn
[import from `utils.ts`][3].

This might be fine in some environments/module formats, but it can cause
unclear errors in the transpiled CommonJS/UMD format (given how Node.js
handles [cycles in module resolution][4]).
(An example error can be found [here][5].)

This commit fixes the problem by moving the code that depends on
`EsmDependencyHost` out of `utils.ts` and into a dedicated file under
`dependencies/`. It also converts the `createDtsDependencyHost()`
function to a class for consistency with the rest of the
`DependencyHost`s.

[1]: https://github.com/angular/angular/blob/18d89c9c8/packages/compiler-cli/ngcc/src/utils.ts#L10
[2]: https://github.com/angular/angular/blob/18d89c9c8/packages/compiler-cli/ngcc/src/dependencies/esm_dependency_host.ts#L10
[3]: https://github.com/angular/angular/blob/18d89c9c8/packages/compiler-cli/ngcc/src/dependencies/dependency_host.ts#L9
[4]: https://nodejs.org/api/modules.html#modules_cycles
[5]: https://circleci.com/gh/angular/angular/577581

mary-poppins commented:
You can preview f567fc3 at https://pr34512-f567fc3.ngbuilds.io/.

gkalpak commented:
Added another commit to fix the CI failure (caused by a cyrcular dependency resolution issue). PTAL

petebacondarwin commented:
LGTM please add "merge" label when you are ready.
